### PR TITLE
Allow compilation and testing in non-Windows environments

### PR DESCRIPTION
**Pull request by jaime-olivares**
The following PR aims to enable full compilation and testing of the entire project on non-Windows machines (tested with macOS). Notice most of the changes are only related to test projects.

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [ ] I have updated API documentation
- [ ] I have included unit tests
- [ ] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- This is the most critical point. Is it necessary to use net6.0-windows and net8.0-windows rather than plain net6.0 and net8.0? This modification works well in macOS
> image

- This is trivial for Windows and enables the compilation on non-Windows environments: 
> image

- This is the unique class that required to replace a literal path with backslahes, with Path.Combine() for cross-platform compatibility:
> image

- Just preventing a warning for the lack of an await in an async method:
> image

- Two test classes and one method required to skip due to incompatibility (WPF, Winforms, FileAttributes.Temporary):
> image

### CODE REVIEW

**Comment by jaime-olivares** [comment]
The fact that the modifications passed the Pull Request pre-checks indicates that they have not affected its operation on Windows.

For testing without issues in non-Windows, it is necessary to use:

`dotnet test --framework:net8.0`

**Review by mrbean-bremen — COMMENTED** [review]
This all looks good to me, thanks! The interesting thing would be to add Linux and maybe Mac builds to the CI. Are you also planning to do this?

**Comment by jaime-olivares** [comment]
Well, we build for Linux and macOS in fo-dicom.Codecs.

Not sure if the main contributors are interested in having that for fo-dicom, but we can collaborate.

**Comment by gofal** [comment]
sure, running the CI actions also on Linux and macOS would be great.

> This is the most critical point. Is it necessary to use net6.0-windows and net8.0-windows rather than plain net6.0 and net8.0? This modification works well in macOS

this has only been, because there is this one testproject, that also tests windows-specific image rendering using system.drawing. That required to build for net6.0-windows. But propably that has been improved in the last years.

**Comment by gofal** [comment]
you added "#IF WINDOWS" to the windows-specific tests.
But tis Platform symbol is only available when a OS-specific target framework moniker is specified:

image
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives

But because you remove them and the targetframeworks now only contain the platform independet "net6.0" and "net8.0", you basically just removed the tests from the testproject. That is why the CI-build is working.

**Comment by gofal** [comment]
a possible solution could be, to create a separate project fo-dicom.Tests.Windows where these two classses are moved to.
and then in CI-scripts on windows both testprojects are executed, while on other platforms only the fo-dicom.tests is executed.

Then the fo-dicom.Tests would not have a reference to fo-dicom.imaging.desktop any more, and you have no need to change the projectfile and add EnableWindowsTargeting any more.

**Comment by mrbean-bremen** [comment]
> a possible solution could be, to create a separate project fo-dicom.Tests.Windows

I agree - that would be better for clean separation of the platforms.

As for the CI - this could be as easy as adding a platform matrix for `runs-on` in the workflow file, we could at least try it.

**Comment by jaime-olivares** [comment]
> a possible solution could be, to create a separate project fo-dicom.Tests.Windows where these two classses are moved to. and then in CI-scripts on windows both testprojects are executed, while on other platforms only the fo-dicom.tests is executed.
> 
> Then the fo-dicom.Tests would not have a reference to fo-dicom.imaging.desktop any more, and you have no need to change the projectfile and add EnableWindowsTargeting any more.

Got it. I will work on that.

File: Documentation/v5/developer/buildnotes.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ Project | Description
 FO-DICOM.Core | This project contains the core logic for handling DICOM files, and working with DICOM services
 Tools/FO-DICOM.Dump | Is is some sample project that can be used to reproduce some isses and debug into the code. The application opens files, shows all tags and values, can anonymize the file and render the image.
 Tests/FO-DICOM.Tests | The xUnit based unit test project
+Tests/FO-DICOM.Tests.Windows | Windows-only unti test project
```
Comment (mrbean-bremen) [inline Documentation/v5/developer/buildnotes.md]: typo: unti

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -18,6 +18,8 @@ jobs:
       run: dotnet build ./FO-DICOM.Core/FO-DICOM.Core.csproj --configuration Release --runtime win-x64
     - name: Test FO-DICOM.Tests
       run: dotnet test ./Tests/FO-DICOM.Tests/FO-DICOM.Tests.csproj --configuration Release --framework net462 --runtime win-x64 --logger:"trx;LogFileName=.\resultsnet462.xml"
+    - name: Test FO-DICOM.Tests
```
Comment (mrbean-bremen) [inline .github/workflows/build.yml]: Test Windows-specific FO-DICOM.Tests?

File: Tests/FO-DICOM.Tests.Windows/WinFormsImageTest.cs [hunk 3]
```diff
@@ -10,13 +10,11 @@
 namespace FellowOakDicom.Tests.Imaging
 {
 
-    [Collection(TestCollections.General)]
+    [Collection("Windows")]
```
Comment (mrbean-bremen) [inline Tests/FO-DICOM.Tests.Windows/WinFormsImageTest.cs:13]: Maybe add an enum value to `TestCollections`?

**Comment by jaime-olivares** [comment]
I have already separated the windows tests as suggested. However, `codecov` is reporting a reduction in coverage.
I even tried to merge the last changes in the development branch, with no luck.

**Comment by mrbean-bremen** [comment]
I guess that the coverage results are not merged together - probably you need [MergeWith](https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/Examples/MSBuild/MergeWith/HowTo.md).
Note that I have no experience here, just looked up the documentation.

**Comment by jaime-olivares** [comment]
@gofal  is this issue related with the recent update of the codecov version?

**Comment by gofal** [comment]
No, Codecov is just a (dumb) tool. Maybe you observed the discussion in the PR about turning warnings into errors? This is a very good example for my arguments in this discussion. Codecov analyzes the code executed during unittests and warns you if after the commit the amount of code executed during unittests decreased. this is a very good hint that you wrote new code where you did not write unittests. or when you removed good tested code by another that is not covered by unittests.
But while this tool is helpful, it is rather dump and we developers are smarter :-)
You moved some tests into another project which is not covered by codecov. So now codecov says that a lot of code now is not tested and wants to prohibit the PR. But we know that the code is still tested, it just moved into another project.

The solution is, to ignore Codecov in this PR here ;-)

**Review by gofal — APPROVED** [review]

**Comment by gofal** [comment]
Ah, i noticed, codecov only takes one resultfile.
the result of test.windows overwrites the result of test.

if you would change the order of execution (first execute the test.windows and then the executeio of test) then codecov will propably read the result of the main test library

**Comment by mrbean-bremen** [comment]
> The solution is, to ignore Codecov in this PR here ;-)

While I agree that it is fine to ignore the check in this PR to get this merged, I don't think we should just leave it at that.
The problem here is that codecov does not merge the results of several tests together, which it is certainly able to do. Being a "dumb tool", it is not able to do so on it's own, but has to be told to. As I wrote, I have no experience with codecov in connection with VS projects, so I'm not sure how to correctly do this, but may have a try some time later.

**Review by mrbean-bremen — APPROVED** [review]

**Comment by gofal** [comment]
> I guess that the coverage results are not merged together - probably you need [MergeWith](https://github.com/coverlet-coverage/coverlet/blob/master/Documentation/Examples/MSBuild/MergeWith/HowTo.md). Note that I have no experience here, just looked up the documentation.

Thanks for the hint. This describes commandline arguments to merge the results. Currently, the codecov in fo-dicom is configured using a configfile `coverlet.runsettings`. So have to take a look at MergeWith how to define it with the settings-file, or if it can be combined, or if we should switch from configfile to commandline.

Would that be a follow-up story to configure codecov for the new test projects, or would you give it a try within this PR?
I would be very fine if we ignore the codecov result for now and reconfigure codecov within a separate story. But I am also ok with doing it here. Jamie?

**Comment by mrbean-bremen** [comment]
> Would that be a follow-up story to configure codecov for the new test projects, or would you give it a try within this PR?

I would see it as a follow-up story, as it is not really part of this PR. Also it may take a bit of trial and error which I'd rather seen in some separate draft PR. I'm fine with merging this as is.

Another follow-up would be to add tests for Linux and possibly macOS in the CI, which should now be possible.

**Comment by mrbean-bremen** [comment]
Going to merge this, I will try to handle result merging in a separate PR, as discussed.

**Comment by mrbean-bremen** [comment]
Just noticed that there are no release notes. @jaime-olivares - can you please add a line in [the change log](https://github.com/fo-dicom/fo-dicom/blob/development/ChangeLog.md)? I think compatibility with other systems is worth an entry.

**Comment by gofal** [comment]
> Just noticed that there are no release notes. @jaime-olivares - can you please add a line in [the change log](https://github.com/fo-dicom/fo-dicom/blob/development/ChangeLog.md)? I think compatibility with other systems is worth an entry.

this pr only affects infrascruture and testing. I think we never had changes in releasenotes for such kind of PRs, since this does not affect the released nuget libraries. I mean, there can be a entry in changelog, but i guess there is no need.

**Comment by jaime-olivares** [comment]
Yes, That’s why I didn’t add any release note

Sent from my iPhone 14
________________________________

**Comment by mrbean-bremen** [comment]
Ok, convinced!

**Pull request merged**